[ews-build.webkit.org] Allow for PRs comprised of cherry-picks to include multiple commits

JonWBedard opened the pull request:
#### 1606ea8ade0f444e257a9e0967faf9b57c7ec31f
<pre>
[ews-build.webkit.org] Allow for PRs comprised of cherry-picks to include multiple commits
<a href="https://bugs.webkit.org/show_bug.cgi?id=254701">https://bugs.webkit.org/show_bug.cgi?id=254701</a>
rdar://107391286

Reviewed by Aakash Jain.

Since cherry-picks nessesarily come from elsewhere in the repository, each cherry-pick
has already been verified to build somewhere. While the previous statement may not apply
to the current branch (a commit may rely on code which does not exist on a branch), it is
exceptionally rare for a cherry-pick failure to be fixed by another cherry-pick. This means
that if a sequence of cherry-picks passes EWS, it is very likely that each individual
component of that sequence would pass EWS in isolation.

To facilitate faster repository management, merge-queue should permit patch series which are
comprised entirely of cherry-picks.

* Tools/CISupport/ews-build/steps.py:
(ValidateSquashed.evaluateCommand): If a PR is comprised entirely of cherry-picks, allow
a PR with multiple commits to be landed. Note that we need to let Canonicalize know how
many commits are going to be landed, because it can no longer safely assume &quot;1&quot;.
(AddReviewerToCommitMessage.doStepIf): If a change is classified as a cherry-pick, we do not
want to add reviewers.
(Canonicalize.number_commits_to_canonicalize): Return the number of commits to be canonicalized.
(Canonicalize.run): Use number_commits_to_canonicalize to determine how many commits to canonicalize.
(Canonicalize.getResultSummary): Ditto.
(UpdatePullRequest.run): Update pull-request which contains multiple commits.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/262340@main">https://commits.webkit.org/262340@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bb870118e87f9f4b13b4b21c6faf6519085ca68d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/1293 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/1329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/1369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/2142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/1162 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/1280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/1381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/1397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/2142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/1304 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/1381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/1369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/1988 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/1381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/1369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/1988 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/1188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/1369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/1988 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/1235 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/1397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/28/builds/1283 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/1174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/1369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/1262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/134 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5646,6 +5666,12 @@ def __init__(self, rebase_enabled=True, **kwargs):
         self.rebase_enabled = rebase_enabled
         self.contributors = {}
 
+    def commit_count(self):
```
is it actually the count of commits or number of commits to canonicalize? If so, better to name it accordingly to avoid confusion.

JonWBedard replied:
Is suppose yes, this is actually the count of the number of commits to canonicalize, which is based on the number of commits in the pull-request.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5502,7 +5519,10 @@ def getResultSummary(self):
+        classification = self.getProperty('classification', [])
+        if not isinstance(classification, list):
+            classification = []
+        return self.getProperty('reviewers_full_names') and ['Cherry-pick'] != classification
```
should it be `'Cherry-pick' not in classification` or you need exact match of both lists?

JonWBedard replied:
We want an exact list match, for now.

Basically, if a PR has multiple types of commits (ie, a `Cherry-pick`, a `Versioning` change and a `Gardening` change), the `classification` list will contain more than just `Cherry-pick`. In those cases, we don't allow multiple commits in the pull-request, our logic for allowing multiple cherry picks in a PR only applies if _all_ changes in the PR are cherry-picks.

aj062 reviewed (approved):
r+ assuming this is tested

JonWBedard committed:
[ews-build.webkit.org] Allow for PRs comprised of cherry-picks to include multiple commits
https://bugs.webkit.org/show_bug.cgi?id=254701
rdar://107391286

Reviewed by Aakash Jain.

Since cherry-picks nessesarily come from elsewhere in the repository, each cherry-pick
has already been verified to build somewhere. While the previous statement may not apply
to the current branch (a commit may rely on code which does not exist on a branch), it is
exceptionally rare for a cherry-pick failure to be fixed by another cherry-pick. This means
that if a sequence of cherry-picks passes EWS, it is very likely that each individual
component of that sequence would pass EWS in isolation.

To facilitate faster repository management, merge-queue should permit patch series which are
comprised entirely of cherry-picks.

* Tools/CISupport/ews-build/steps.py:
(ValidateSquashed.evaluateCommand): If a PR is comprised entirely of cherry-picks, allow
a PR with multiple commits to be landed. Note that we need to let Canonicalize know how
many commits are going to be landed, because it can no longer safely assume "1".
(AddReviewerToCommitMessage.doStepIf): If a change is classified as a cherry-pick, we do not
want to add reviewers.
(Canonicalize.number_commits_to_canonicalize): Return the number of commits to be canonicalized.
(Canonicalize.run): Use number_commits_to_canonicalize to determine how many commits to canonicalize.
(Canonicalize.getResultSummary): Ditto.
(UpdatePullRequest.run): Update pull-request which contains multiple commits.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: https://commits.webkit.org/262340@main
